Rewrite both shift registers

nulldg opened the pull request:
The SIPO and PISO are cleaner now, DRY violations are gone. The SIPO is now more efficient because it's no longer wasting resources storing and manipulating redundant information. Of course, I've also enabled changing the number of bits (since not every shift register IC is 8-bit)

I've also added ``defaultBitCount()`` to ``ChipElm.java`` to allow elements to use a different default bit count: the number of bits to use when creating a new element, or when the bit count has been omitted in the dump for some reason (eg loading old circuits from a time when "# of bits" was hard coded). There used to be a hard-coded exception for the Ring Counter to give it a default of 10, but the shift registers now need an exception too, because the default bit count of the shift registers needs to be 8 to not break old circuits, whereas the default is 4.

Although I've unit tested these shift registers and tested loading an old dump, I don't have a good circuit to properly test them with.

pfalstad commented:
looks great.  I don't have any old circuits that uses these.  One thing I noticed is that, like the old version, it doesn't save its state when exporting/important.  For the Sipo you just need to set state=true for all the output pins.  For Piso it would be more work.

nulldg commented:
The states are now exported and imported as integers. I've ensured that even if bits are missing from the data, the registers will still load successfully.

nulldg commented:
Hmm, I'm concerned the pins are backwards. Q1 feeds into Q2, which feeds into Q3. In Edward's implementation, the latches feed the wrong way round. This would be a rather trivial fix, though.

nulldg commented:
The obvious cosmetic solution (renaming the pins) does has a problem, though: Asymmetry between the SIPO and PISO shift registers.
![image](https://user-images.githubusercontent.com/55410457/121901460-52fda680-cd1e-11eb-8fdd-a362aa256796.png)
This isn't really a big deal, but I wish I could flip the PISO (so data moves towards the output, just as the SIPO's data moves away from the input) without breaking old circuits.

nulldg commented:
The shift registers now work the same as the 74HC164 and 74HC165.
But unlike the 74HC165, the PISO has an extra output register. This isn't obvious. This new labeling should hopefully allude to this.
But by fixing the labelling, we've now got a cosmetic problem with symmetry, where the PISO and SIPO are both oriented in opposite directions.
![image](https://user-images.githubusercontent.com/55410457/121909151-94458480-cd25-11eb-9fbb-4a8ef3399c7c.png)
The SIPO is perfect, so if this is enough of a problem, we should change the PISO.

pfalstad commented:
The PISO pins should be renamed so D0 is on the left, shouldn't they?

The bigger problem is that the PISO doesn't work like the 74HC165.  If I'm reading this datasheet correctly, whatever you load into D7 is immediately reflected in the output (Q7).

pfalstad commented:
Also the code to dump/undump an array of bits should be shared by moving it into ChipElm.

Using StringBuilder is fine but it's basically the same as string concatenation in GWT, so I wouldn't go out of my way to use it.  Here is how it is implemented: https://github.com/gwtproject/gwt/blob/master/user/super/com/google/gwt/emul/java/lang/StringBuilder.java

nulldg commented:
> The PISO pins should be renamed so D0 is on the left, shouldn't they?

I compared the simulated behavior against the datasheet's examples and it seems to check out. Of course, I could be wrong. I'll test it again in a sec.


> The bigger problem is that the PISO doesn't work like the 74HC165

Yeah, that's what I said. This PISO has an extra output register for some reason, but I don't think we can change this behavior now.


> Using StringBuilder is fine but it's basically the same as string concatenation in GWT

Wow, that's disappointing. StringBuilders are normally significantly faster than standard concatenation. I could have sworn javascript had buffers or something which would have let the GWT developers retain the performance boosts.

pfalstad commented:
oh extra output REGISTER.  gotcha.

nulldg commented:
> Also the code to dump/undump an array of bits should be shared by moving it into ChipElm.

I'm having difficulty with this. The code may look duplicate, but there are cardinal differences between the import/export functions of both shift registers. Without anonymous functions in Java, these differences are difficult to reconcile.
When exporting: The PISO needs to read its data as a circular array (not a direct index!), and the SIPO reads its data directly from its pins' ``value``s.
When importing: The PISO reads bits into the data array, but the SIPO needs to assign the values to its pins' ``value``.

pfalstad commented:
You can leave it like it is if you want and I'll take a look at it.  I will probably copy the bits into a temporary array.

nulldg commented:
I worry that that'll end up resulting in more duplicate code.

nulldg commented:
Thinking more about the PISO, maybe it won't be necessary to treat the data like a circular array, considering how the PISO shifts destructively. Data prior to the circular array's index is just garbage anyway. As long as enough zero-padding is added to the end, this would simplify it.

pfalstad commented:
Either D0 should be on the left, or the output should be renamed Q-1 (which would be weird).  If we didn't have the extra output register, then the output would be named after the input right next to it, since it is the same.  But since we do have the extra register, it should be one larger/smaller.

nulldg commented:
I've been thinking more about this problem.

The pins are 0-indexed numbers referring to the index of the internal shift register it's connected to. Pins prefixed with D are inputs, and pins prefixed with Q are outputs.

D0 is the most significant bit which means it's the first to go out, and Dh is the least significant digit, so it's the last to go out. The reason the final pin is labeled Q8 (for an 8-bit SR) is because it represents the final output register, after D7.

This is the 74HC165. Notice that when clocking out, the first bit to go out is H (which would correspond to pin D7) and A is the last (which would correspond to D0). In other words, the data moves from A to B to C to D, etc until it reaches the output at the end.
![image](https://user-images.githubusercontent.com/55410457/122312701-adf4f080-cf0c-11eb-96ab-3d180899f770.png)
![image](https://user-images.githubusercontent.com/55410457/122313779-bc440c00-cf0e-11eb-9183-b76e56a7a41c.png)

For circuitjs's PISO, I know the data is physically going in the wrong direction away from the output. Believe me, I genuinely hate it. It's always been incredibly unintuitive.
I would love to simply move the pins around so it makes sense and works the same way the SIPO does. But we cannot move the data pins nor can we expose the final output register without breaking old circuits.
If we just change the labels and not the pins, then the data would _still_ being going in the wrong direction AND the shift register would be technically inaccurate.
Unfortunately, it was Edward's choice to make the data go backwards away from the output, and sadly we cannot do anything about that unless we add a flag for it.

pfalstad commented:
I'm confused.  It looks like the bits are going the right way (to the right).  Here is a test circuit.

https://tinyurl.com/yflqttv5

Click load, and you will get 10000111 on the output.

nulldg commented:
Whoops, I think I confused myself by mistaking branches. Ignore the points about the data going the wrong direction, sorry!

nulldg commented:
Yeah, Edward still got the indexes the wrong way round, likely because he was unnecessarily manipulating the PISO's state in bitwise, but the visual direction actually is correct.
Everything is good now.

nulldg commented:
Do you think we should add a SER pin? It's often used to combine several PISO shift registers to increase the number of bits, but it's useful for other applications too (eg using the parallel-in-serial-out as a serial-in-serial-out shift register, or providing overflow data)

pfalstad commented:
If you want, but then won't we have to remove the extra register?  We could certainly do that with an extra flag bit.

nulldg commented:
SER would fill the gap at Q0 as the data is shifted up. Right now we're just filling the gap implicitly with 0s. Unless I'm mistaken, I don't think the extra register would affect this.

pfalstad commented:
Yes, adding SER and/or removing the hidden register would be good if you're willing.

nulldg commented:
How should we deal with the hidden register? Should all new shift registers simply omit it? Or should there be an option for enabling/disabling it?

pfalstad commented:
omit it I think.  there's no reason to enable it.  the only reason I would add the option is to make it more obvious that there is a difference between old shift registers and new ones.  but I think that is clear by looking at the output, which will be labelled differently, right?

nulldg commented:
Yep, it will be.

pfalstad commented:
I didn't merge this because it sounds like you're still working on it, right?

nulldg commented:
Yep, I still would like to add the SER pin. Sorry for the delay, I've been quite busy recently.

pfalstad commented:
no rush at all, don't worry.

nulldg commented:
Looks like I'm done!
As the commit name suggests, I've added the SER pin, removed the hidden register, and simplified `writeBits` since the extra arguments are not needed anymore. Although I've done my best to preserve the old behavior for old circuits, I think more testing is needed.
I would use an old circuit of mine, but none of my circuits use these shift registers because of the limitations circuitjs's shift registers had. (I used subcircuits instead, but subcircuits can be rather slow for large circuits)

pfalstad commented:
Excellent!!  Looks great!  I don't have any old test circuits either, unfortunately.

nulldg commented:
I'll try make a digital circuit using the master branch, and then import it using this branch to make sure there aren't any bugs I didn't catch by simply unit testing it.

nulldg commented:
Good news, it works. This is the old circuit I tested it with. The circuit takes 4 milliseconds to start working because I wanted to make sure the registers' initial conditions are preserved across dumps and the version change.
> $ 1 0.000005 3.3115451958692312 50 5 50 5e-11
405 992 128 1152 128 0 8 5
186 464 624 608 624 0
186 464 448 608 448 0
186 464 272 608 272 0
186 464 96 608 96 0
186 464 800 608 800 0
L 528 64 528 32 0 0 false 5 0
L 560 64 560 32 0 1 false 5 0
L 592 64 592 32 0 0 false 5 0
L 624 64 624 32 0 0 false 5 0
L 656 64 656 32 0 1 false 5 0
L 688 64 688 32 0 1 false 5 0
L 720 64 720 32 0 0 false 5 0
L 752 64 752 32 0 0 false 5 0
L 752 240 752 208 0 0 false 5 0
L 720 240 720 208 0 1 false 5 0
L 688 240 688 208 0 0 false 5 0
L 656 240 656 208 0 1 false 5 0
L 624 240 624 208 0 0 false 5 0
L 592 240 592 208 0 1 false 5 0
L 560 240 560 208 0 0 false 5 0
L 528 240 528 208 0 1 false 5 0
L 752 416 752 384 0 0 false 5 0
L 720 416 720 384 0 0 false 5 0
L 688 416 688 384 0 1 false 5 0
L 656 416 656 384 0 1 false 5 0
L 624 416 624 384 0 0 false 5 0
L 592 416 592 384 0 1 false 5 0
L 560 416 560 384 0 1 false 5 0
L 528 416 528 384 0 1 false 5 0
L 528 592 528 560 0 1 false 5 0
L 560 592 560 560 0 0 false 5 0
L 592 592 592 560 0 1 false 5 0
L 624 592 624 560 0 0 false 5 0
L 656 592 656 560 0 1 false 5 0
L 688 592 688 560 0 0 false 5 0
L 720 592 720 560 0 1 false 5 0
L 752 592 752 560 0 0 false 5 0
L 528 768 528 736 0 1 false 5 0
L 560 768 560 736 0 0 false 5 0
L 592 768 592 736 0 1 false 5 0
L 624 768 624 736 0 0 false 5 0
L 656 768 656 736 0 1 false 5 0
L 688 768 688 736 0 1 false 5 0
L 720 768 720 736 0 0 false 5 0
L 752 768 752 736 0 0 false 5 0
r 992 128 912 128 0 470
r 992 160 912 160 0 470
r 992 192 912 192 0 470
r 992 224 912 224 0 470
w 816 128 912 128 0
w 880 160 848 192 0
w 848 192 848 272 0
w 848 272 816 304 0
w 880 192 864 208 0
w 864 208 864 448 0
w 864 448 832 480 0
w 912 160 880 160 0
w 816 480 832 480 0
w 880 192 912 192 0
w 896 224 880 240 0
w 880 240 880 624 0
w 880 624 848 656 0
w 848 656 816 656 0
w 864 832 896 800 0
w 864 832 816 832 0
w 896 224 912 224 0
r 992 256 912 256 0 470
w 912 256 896 272 0
w 896 272 896 800 0
163 960 432 1072 432 1030 10 0 0 0 0 0 0 0 0 0 0
g 1248 496 1248 512 0 0
w 1344 496 1344 896 0
w 1312 928 352 928 0
w 352 928 384 896 0
w 464 864 416 864 0
w 464 832 352 832 0
w 384 864 384 720 0
w 320 800 320 624 0
w 464 688 416 688 0
w 464 656 352 656 0
w 384 720 384 544 0
w 320 624 320 448 0
w 464 512 416 512 0
w 464 480 352 480 0
w 384 544 384 368 0
w 320 448 320 272 0
w 464 336 416 336 0
w 464 304 352 304 0
w 384 368 384 192 0
w 320 272 320 96 0
w 464 160 416 160 0
w 464 128 352 128 0
w 384 896 384 864 0
w 1312 928 1344 896 0
w 1312 464 1344 496 0
w 352 832 320 800 0
w 352 656 320 624 0
w 352 480 320 448 0
w 352 304 320 272 0
w 352 128 320 96 0
w 928 352 928 16 0
w 928 16 896 -16 0
w 896 -16 352 -16 0
w 352 -16 320 16 0
w 416 160 384 192 0
w 384 368 416 336 0
w 416 512 384 544 0
w 384 720 416 688 0
w 384 896 416 864 0
R 352 928 192 928 1 2 2000 2.5 2.5 0 0.5
L 288 -16 208 -16 0 0 true 5 0
d 992 400 992 352 2 default
d 288 -16 352 -16 2 default
I 1024 400 1024 288 0 0.5 5
I 1056 400 1056 288 0 0.5 5
I 1088 400 1088 288 0 0.5 5
I 1120 400 1120 288 0 0.5 5
I 1248 400 1248 288 0 0.5 5
I 1216 400 1216 288 0 0.5 5
I 1184 400 1184 288 0 0.5 5
I 1152 400 1152 288 0 0.5 5
l 928 416 928 496 0 0.1 0 0
g 928 560 928 608 0 0
r 928 560 928 496 0 1000
r 928 352 928 416 0 390
w 928 352 992 352 0
w 928 496 992 496 0
w 320 16 320 96 0
x 1274 359 1393 395 4 8 Extra\spin\sto\saccount\sfor\sthe\\nhidden\soutput\sregister\sin\sthe\sPISOs.\\nRemove\sthis\spin\sif\sthe\sextra\\nregister\sis\sremoved/disabled.


Here is an updated version which is slightly cleaner because it no longer has the hidden register.
> $ 1 0.000005 3.046768661252054 50 5 50 5e-11
405 992 128 1152 128 0 8 5
L 528 64 528 32 0 0 false 5 0
L 560 64 560 32 0 1 false 5 0
L 592 64 592 32 0 0 false 5 0
L 624 64 624 32 0 0 false 5 0
L 656 64 656 32 0 1 false 5 0
L 688 64 688 32 0 1 false 5 0
L 720 64 720 32 0 0 false 5 0
L 752 64 752 32 0 0 false 5 0
L 752 240 752 208 0 0 false 5 0
L 720 240 720 208 0 1 false 5 0
L 688 240 688 208 0 0 false 5 0
L 656 240 656 208 0 1 false 5 0
L 624 240 624 208 0 0 false 5 0
L 592 240 592 208 0 1 false 5 0
L 560 240 560 208 0 0 false 5 0
L 528 240 528 208 0 1 false 5 0
L 752 416 752 384 0 0 false 5 0
L 720 416 720 384 0 0 false 5 0
L 688 416 688 384 0 1 false 5 0
L 656 416 656 384 0 1 false 5 0
L 624 416 624 384 0 0 false 5 0
L 592 416 592 384 0 1 false 5 0
L 560 416 560 384 0 1 false 5 0
L 528 416 528 384 0 1 false 5 0
L 528 592 528 560 0 1 false 5 0
L 560 592 560 560 0 0 false 5 0
L 592 592 592 560 0 1 false 5 0
L 624 592 624 560 0 0 false 5 0
L 656 592 656 560 0 1 false 5 0
L 688 592 688 560 0 0 false 5 0
L 720 592 720 560 0 1 false 5 0
L 752 592 752 560 0 0 false 5 0
L 528 768 528 736 0 1 false 5 0
L 560 768 560 736 0 0 false 5 0
L 592 768 592 736 0 1 false 5 0
L 624 768 624 736 0 0 false 5 0
L 656 768 656 736 0 1 false 5 0
L 688 768 688 736 0 1 false 5 0
L 720 768 720 736 0 0 false 5 0
L 752 768 752 736 0 0 false 5 0
r 992 128 912 128 0 470
r 992 160 912 160 0 470
r 992 192 912 192 0 470
r 992 224 912 224 0 470
w 816 128 912 128 0
w 880 160 848 192 0
w 848 192 848 272 0
w 848 272 816 304 0
w 880 192 864 208 0
w 864 208 864 448 0
w 864 448 832 480 0
w 912 160 880 160 0
w 816 480 832 480 0
w 880 192 912 192 0
w 896 224 880 240 0
w 880 240 880 624 0
w 880 624 848 656 0
w 848 656 816 656 0
w 864 832 896 800 0
w 864 832 816 832 0
w 896 224 912 224 0
r 992 256 912 256 0 470
w 912 256 896 272 0
w 896 272 896 800 0
163 960 432 1072 432 1030 9 0 0 0 0 0 0 0 0 0
g 1216 496 1216 512 0 0
w 1312 496 1312 896 0
w 1280 928 352 928 0
w 352 928 384 896 0
w 464 864 416 864 0
w 464 832 352 832 0
w 384 864 384 720 0
w 320 800 320 624 0
w 464 688 416 688 0
w 464 656 352 656 0
w 384 720 384 544 0
w 320 624 320 448 0
w 464 512 416 512 0
w 464 480 352 480 0
w 384 544 384 368 0
w 320 448 320 272 0
w 464 336 416 336 0
w 464 304 352 304 0
w 384 368 384 192 0
w 320 272 320 96 0
w 464 160 416 160 0
w 464 128 352 128 0
w 384 896 384 864 0
w 1280 928 1312 896 0
w 1280 464 1312 496 0
w 352 832 320 800 0
w 352 656 320 624 0
w 352 480 320 448 0
w 352 304 320 272 0
w 352 128 320 96 0
w 928 352 928 16 0
w 928 16 896 -16 0
w 896 -16 352 -16 0
w 352 -16 320 16 0
w 416 160 384 192 0
w 384 368 416 336 0
w 416 512 384 544 0
w 384 720 416 688 0
w 384 896 416 864 0
R 352 928 192 928 1 2 2000 2.5 2.5 0 0.5
L 288 -16 208 -16 0 0 true 5 0
d 992 400 992 352 2 default
d 288 -16 352 -16 2 default
I 1024 400 1024 288 0 0.5 5
I 1056 400 1056 288 0 0.5 5
I 1088 400 1088 288 0 0.5 5
I 1120 400 1120 288 0 0.5 5
I 1248 400 1248 288 0 0.5 5
I 1216 400 1216 288 0 0.5 5
I 1184 400 1184 288 0 0.5 5
I 1152 400 1152 288 0 0.5 5
l 928 416 928 496 0 0.1 0 0
g 928 560 928 608 0 0
r 928 560 928 496 0 1000
r 928 352 928 416 0 390
w 928 352 992 352 0
w 928 496 992 496 0
w 320 16 320 96 0
186 464 96 560 96 2 8 1
186 464 272 560 272 2 8 2
186 464 448 560 448 2 8 3
186 464 624 560 624 2 8 2
186 464 800 560 800 2 8 2
g 464 96 448 96 0 0
g 464 272 448 272 0 0
g 464 448 448 448 0 0
g 464 624 448 624 0 0
g 464 800 448 800 0 0
